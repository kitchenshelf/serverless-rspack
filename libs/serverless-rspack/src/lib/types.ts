import { RspackOptions } from "@rspack/core";

export interface Configuration extends RspackOptions {
  packagePath: string;

  keepOutputDirectory?: boolean;
  outputWorkFolder?: string;
  outputBuildFolder?: string;

  skipBuild?: boolean;
  esm?: boolean;
}
