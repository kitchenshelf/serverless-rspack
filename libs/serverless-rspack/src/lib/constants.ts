export const SERVERLESS_FOLDER = '.serverless';
export const WORK_FOLDER = '.rspack';
