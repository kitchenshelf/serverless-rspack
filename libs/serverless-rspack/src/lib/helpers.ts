import type { FunctionDefinitionHandler } from 'serverless';

export const humanSize = (size: number) => {
  if (size === 0) {
    return '0.00 B';
  }
  if (size >= Number.MAX_SAFE_INTEGER) {
    return 'MSI';
  }
  const exponent = Math.floor(Math.log(size) / Math.log(1024));
  const sanitized = (size / 1024 ** exponent).toFixed(2);

  return `${sanitized} ${['B', 'KB', 'MB', 'GB', 'TB'][exponent]}`;
};

/**
 * Checks if the runtime for the given function is nodejs.
 * If the runtime is not set , checks the global runtime.
 * @param {FunctionDefinitionHandler} func the function to be checked
 * @returns {boolean} true if the function/global runtime is nodejs; false, otherwise
 */
export function isNodeFunction(
  func: FunctionDefinitionHandler,
  providerRuntime: string | undefined
): boolean {
  const runtime = (func.runtime || providerRuntime) ?? '';

  return typeof runtime === 'string' && runtime.startsWith('node');
}

export function determineFileParts(handlerFile: string) {
  const regex = /^(.*)\/([^/]+)$/;
  const result = regex.exec(handlerFile);

  const filePath = result?.[1] ?? '';
  const fileName = result?.[2] ?? handlerFile;
  return { filePath, fileName };
}

export function enabledViaSimpleConfig(field: unknown): field is boolean {
  return typeof field === 'boolean' && field === true;
}

export function enabledViaConfigObject<T extends { enable?: boolean }>(
  field: T
): field is T {
  return (
    typeof field === 'object' &&
    field !== null &&
    (field.enable === true || field.enable === undefined)
  );
}
