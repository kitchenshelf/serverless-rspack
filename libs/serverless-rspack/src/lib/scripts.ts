import type { RspackServerlessPlugin } from './serverless-rspack.js';
import { execSync } from 'node:child_process';

export async function scripts(this: RspackServerlessPlugin) {
  if (this.pluginOptions.scripts && this.pluginOptions.scripts.length > 0) {
    this.log.verbose(
      `[Scripts] Running ${this.pluginOptions.scripts.length} scripts`
    );
    const startScripts = Date.now();
    for (const [index, script] of this.pluginOptions.scripts.entries()) {
      const startZip = Date.now();
      try {
        execSync(script, {
          cwd: this.serviceDirPath,
          stdio: this.options.verbose ? 'inherit' : 'ignore',
          env: {
            ...process.env,
            KS_SERVICE_DIR: this.serviceDirPath,
            KS_BUILD_OUTPUT_FOLDER: this.buildOutputFolder,
            KS_PACKAGE_OUTPUT_FOLDER: this.packageOutputFolder,
          },
        });
        this.log.verbose(
          `[Performance] Script ${index + 1} of ${
            this.pluginOptions.scripts.length
          } [${Date.now() - startZip} ms]`
        );
      } catch (error) {
        throw new this.serverless.classes.Error(
          `Failed to execute script: ${script}`
        );
      }
    }
    this.log.verbose(
      `[Performance] Scripts total execution time for service ${
        this.serverless.service.service
      } [${Date.now() - startScripts} ms]`
    );
  }
}
