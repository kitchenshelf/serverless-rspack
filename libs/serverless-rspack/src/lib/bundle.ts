import { RspackOptions, rspack } from '@rspack/core';
import { writeFileSync } from 'node:fs';
import path from 'node:path';
import { defaultConfig } from './helpers.js';
import type { RspackServerlessPlugin } from './serverless-rspack.js';

export async function bundle(
  this: RspackServerlessPlugin,
  entries: RspackOptions['entry']
) {
  let config: RspackOptions;
  if (this.providedRspackConfig) {
    config = {
      ...this.providedRspackConfig,
      entry: entries,
      output: {
        ...this.providedRspackConfig.output,
        path: this.buildOutputFolderPath,
      },
    };
  } else {
    config = defaultConfig(
      entries,
      this.pluginConfig,
      this.buildOutputFolderPath
    );
  }
  this.log.verbose('Bundling with config: ', config);
  const startPack = Date.now();

  return new Promise((resolve) => {
    rspack(config, (x, y) => {
      if (this.pluginConfig.stats) {
        const c = y?.toJson();
        try {
          writeFileSync(
            path.join(this.buildOutputFolderPath, 'stats.json'),
            JSON.stringify(c)
          );
        } catch (error) {
          this.log?.error('Failed to write stats file: ', error);
        }
      }
      this.log.verbose(
        `[PERFORMANCE] Bundle service ${this.serverless.service.service} [${
          Date.now() - startPack
        } ms]`
      );
      resolve('Success!'); // Yay! Everything went well!
    });
  });
}
