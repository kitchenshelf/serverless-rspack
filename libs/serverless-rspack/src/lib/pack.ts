import archiver from 'archiver';
import fs from 'node:fs';
import path from 'node:path';

import { humanSize } from './helpers.js';
import type { RspackServerlessPlugin } from './serverless-rspack.js';

export async function pack(this: RspackServerlessPlugin) {
  if (!this.functionEntries) {
    return;
  }
  const pMap = (await import('p-map')).default;
  const zipMapper = async (name: string) => {
    const loadedFunc = this.serverless.service.getFunction(name);
    const zipName = `${name}.zip`;

    const artifactPath = path.join(
      this.serviceDirPath,
      this.packageOutputPath,
      zipName
    );
    try {
      const startZip = Date.now();

      await zipDirectory(
        this.serviceDirPath + '/' + this.buildOutputFolder + '/' + name,
        artifactPath
      );
      const { size } = fs.statSync(artifactPath);

      this.log.verbose(
        `[PERFORMANCE] Pack service ${this.serverless.service.service}: ${
          loadedFunc.name
        } -  ${humanSize(size)} [${Date.now() - startZip} ms]`
      );
    } catch (error) {
      this.log.error(error);
    }
    loadedFunc.package = {
      artifact: artifactPath,
    };
  };

  this.log.verbose(
    `[PERFORMANCE] Pack service ${this.serverless.service.service} with concurrency: [${this.pluginConfig.zipConcurrency}] `
  );

  await pMap(
    Object.entries(this.functionEntries).map((x) => x[0]),
    zipMapper,
    {
      concurrency: this.pluginConfig.zipConcurrency,
    }
  );
}

/**
 * From: https://stackoverflow.com/a/51518100
 * @param {String} sourceDir: /some/folder/to/compress
 * @param {String} outPath: /path/to/created.zip
 * @returns {Promise}
 */
function zipDirectory(sourceDir: string, outPath: string) {
  const archive = archiver('zip', { zlib: { level: 9 } });
  const stream = fs.createWriteStream(outPath);

  return new Promise((resolve, reject) => {
    archive
      .directory(sourceDir, false)
      .on('error', (err) => reject(err))
      .pipe(stream);

    stream.on('close', () => resolve('success'));
    void archive.finalize();
  });
}
