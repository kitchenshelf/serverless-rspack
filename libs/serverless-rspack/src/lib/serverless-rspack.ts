import { RspackOptions } from '@rspack/core';
import { assert } from 'node:console';
import { readdirSync } from 'node:fs';
import { rm } from 'node:fs/promises';
import path from 'node:path';
import type Serverless from 'serverless';
import type ServerlessPlugin from 'serverless/classes/Plugin';
import { bundle } from './bundle.js';
import { SERVERLESS_FOLDER, WORK_FOLDER } from './constants.js';
import { pack } from './pack.js';
import {
  PluginOptions,
  PluginOptionsSchema,
  RsPackFunctionDefinitionHandler,
} from './types.js';

export class RspackServerlessPlugin implements ServerlessPlugin {
  serviceDirPath: string;
  buildOutputFolder: string;
  buildOutputFolderPath: string;
  packageOutputPath: string;

  log: ServerlessPlugin.Logging['log'];
  serverless: Serverless;
  options: Serverless.Options;

  pluginOptions!: PluginOptions;
  providedRspackConfig: RspackOptions | undefined;

  functionEntries: RspackOptions['entry'] | undefined;

  hooks: ServerlessPlugin.Hooks;

  timings = new Map<string, number>();

  #bundle = bundle.bind(this);
  #pack = pack.bind(this);

  constructor(
    serverless: Serverless,
    options: Serverless.Options,
    logging: ServerlessPlugin.Logging
  ) {
    assert(logging, 'Please use serverless V4');

    serverless.configSchemaHandler.defineFunctionProperties('aws', {
      properties: {
        rspack: { type: 'boolean' },
      },
    });

    this.serverless = serverless;
    this.options = options;
    this.log = logging.log;
    this.serviceDirPath = this.serverless.config.serviceDir;
    this.packageOutputPath = SERVERLESS_FOLDER;
    this.buildOutputFolder = WORK_FOLDER;
    this.buildOutputFolderPath = path.join(
      this.serviceDirPath,
      this.buildOutputFolder
    );

    this.hooks = {
      initialize: async () => {
        await this.init();
      },
      // 'before:run:run': async () => {
      // this.log.verbose('before:run:run');
      // await this.#bundle(this.functionEntries);
      // },
      // 'before:offline:start': () => {
      //   this.log.verbose('before:offline:start');
      // },
      // 'before:offline:start:init': () => {
      //   this.log.verbose('before:offline:start:init');
      // },
      'before:package:createDeploymentArtifacts': async () => {
        this.log.verbose('before:package:createDeploymentArtifacts');
        this.timings.set(
          'before:package:createDeploymentArtifacts',
          Date.now()
        );

        if (
          !this.functionEntries ||
          Object.entries(this.functionEntries).length === 0
        ) {
          throw new this.serverless.classes.Error(
            `No function entries provided - try running in verbose mode to see expected entries`
          );
        }
        await this.#bundle(this.functionEntries);
        await this.#pack();
      },
      'after:package:createDeploymentArtifacts': async () => {
        this.log.verbose('after:package:createDeploymentArtifacts');
        await this.cleanup();
        this.log.verbose(
          `[Performance] Hook createDeploymentArtifacts ${
            this.serverless.service.service
          } [${
            Date.now() -
            this.timings.get('before:package:createDeploymentArtifacts')!
          } ms]`
        );
      },
      // 'before:deploy:function:packageFunction': () => {
      //   this.log.verbose('after:deploy:function:packageFunction');
      // },
      // 'after:deploy:function:packageFunction': () => {
      //   this.log.verbose('after:deploy:function:packageFunction');
      // },
      // 'before:invoke:local:invoke': async () => {
      //   this.log.verbose('before:invoke:local:invoke');
      //   this.log.verbose(this.options);
      //   // @ts-ignore
      //   const invokeFunc: string =
      //     this.serverless.processedInput.options.function;
      //   this.log.verbose(invokeFunc);
      //   await this.#bundle({
      //     [invokeFunc]: (this.functionEntries as any)[invokeFunc],
      //   });

      //   this.serviceDirPath = this.buildOutputFolderPath;
      //   this.serverless.config.servicePath = this.buildOutputFolderPath;
      //   process.chdir(this.serviceDirPath);
      // },
      // 'after:invoke:local:invoke': () => {
      //   this.log.verbose('after:invoke:local:invoke');
      // },
    };
  }

  private async init() {
    this.pluginOptions = this.getPluginOptions();

    if (this.pluginOptions.config?.path) {
      const configPath = path.join(
        this.serviceDirPath,
        this.pluginOptions.config.path
      );
      if (!this.serverless.utils.fileExistsSync(configPath)) {
        throw new this.serverless.classes.Error(
          `Rspack config does not exist at path: ${configPath}`
        );
      }
      const configFn = (await import(configPath)).default;

      if (typeof configFn !== 'function') {
        throw new this.serverless.classes.Error(
          `Config located at ${configPath} does not return a function. See for reference: https://github.com/kitchenshelf/serverless-rspack/blob/main/README.md#config-file`
        );
      }

      this.providedRspackConfig = configFn(this.serverless);
    }

    const functions = this.serverless.service.getAllFunctions();
    this.functionEntries = this.buildFunctionEntries(functions);
    this.log.verbose('Function Entries:', this.functionEntries);
  }

  private buildFunctionEntries(functions: string[]) {
    this.log.verbose('Building function entries for: ', functions);
    let entries = {};

    functions.forEach((functionName) => {
      const functionDefinitionHandler = this.serverless.service.getFunction(
        functionName
      ) as RsPackFunctionDefinitionHandler;
      if (
        functionDefinitionHandler.rspack ||
        this.isNodeFunction(functionDefinitionHandler)
      ) {
        // TODO: support container images
        const entry = this.getEntryForFunction(
          functionName,
          functionDefinitionHandler as Serverless.FunctionDefinitionHandler
        );
        entries = {
          ...entries,
          ...entry,
        };
      }
    });
    return entries;
  }

  private getEntryForFunction(
    name: string,
    serverlessFunction: Serverless.FunctionDefinitionHandler
  ) {
    const handler = serverlessFunction.handler;
    this.log.verbose(
      `Processing function ${name} with provided handler ${handler}`
    );

    const handlerFile = this.getHandlerFile(handler);

    const { filePath, fileName } = this.determineFileParts(handlerFile);
    const safeFilePath = filePath ? '/' + filePath + '/' : '/';

    const files = readdirSync(`./${filePath}`);

    const file = files.find((file) => {
      return file.startsWith(fileName);
    });

    if (!file) {
      throw new this.serverless.classes.Error(
        `Unable to find file [${fileName}] in path: [./${filePath}]`
      );
    }
    const ext = path.extname(file);

    this.log.verbose(
      `Determined: filePath: [${safeFilePath}] - fileName: [${fileName}] - ext: [${ext}]`
    );
    const outputExtension = this.isESM() ? 'mjs' : 'js';

    return {
      [name]: {
        import: `./${handlerFile}${ext}`,
        filename: `[name]${safeFilePath}${fileName}.${outputExtension}`,
      },
    };
  }

  private isESM() {
    return (
      this.providedRspackConfig?.experiments?.outputModule ||
      this.pluginOptions.esm
    );
  }

  private determineFileParts(handlerFile: string) {
    const regex = /^(.*)\/([^/]+)$/;
    const result = regex.exec(handlerFile);

    const filePath = result?.[1] ?? '';
    const fileName = result?.[2] ?? handlerFile;
    return { filePath, fileName };
  }

  private getHandlerFile(handler: string) {
    // Check if handler is a well-formed path based handler.
    const handlerEntry = /(.*)\..*?$/.exec(handler);
    if (handlerEntry) {
      return handlerEntry[1];
    }
    throw new this.serverless.classes.Error(`malformed handler: ${handler}`);
  }

  /**
   * Checks if the runtime for the given function is nodejs.
   * If the runtime is not set , checks the global runtime.
   * @param {Serverless.FunctionDefinitionHandler} func the function to be checked
   * @returns {boolean} true if the function/global runtime is nodejs; false, otherwise
   */
  private isNodeFunction(func: Serverless.FunctionDefinitionHandler): boolean {
    const runtime = func.runtime || this.serverless.service.provider.runtime;

    return typeof runtime === 'string' && runtime.startsWith('node');
  }

  private async cleanup(): Promise<void> {
    if (!this.pluginOptions.keepOutputDirectory) {
      await rm(path.join(this.buildOutputFolderPath), { recursive: true });
    }
  }

  private getPluginOptions() {
    return PluginOptionsSchema.parse(
      this.serverless.service.custom?.['rspack'] ?? {}
    );
  }
}
