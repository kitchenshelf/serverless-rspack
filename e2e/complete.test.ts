import fs from 'fs-extra';
import path from 'path';
import { validateLambdaFunction } from './lib/utilities';

describe('Lambda Function Deployments', () => {
  let cloudformation: any;
  let testArtifactPath: string;
  const tempAssertDir = path.join('./assert-complete' + Date.now());

  beforeAll(() => {
    testArtifactPath = path.resolve(
      __dirname,
      '../examples/complete/.serverless'
    );

    cloudformation = require(path.join(
      testArtifactPath,
      'cloudformation-template-update-stack.json'
    ));
  });

  afterEach(async () => {
    // Clean extracted files between tests
    await fs.emptyDir(tempAssertDir);
  });

  afterAll(() => {
    fs.rmSync(tempAssertDir, { recursive: true });
  });

  test('Top level cloudformation', async () => {
    expect(cloudformation.AWSTemplateFormatVersion).toMatchSnapshot();
  });

  test('Correct zips are outputted', async () => {
    const zipFiles = fs
      .readdirSync(testArtifactPath)
      .filter((file) => file.endsWith('.zip'));

    const expectedZipFiles = [
      'App1.zip',
      'app2.zip', //Handled and packaged by serverless (python)
      'app3.zip',
      'app4.zip',
      'app5.zip',
      'app6.zip', // Handled and packaged by serverless (rspack false)
    ];

    expect(zipFiles).toEqual(expect.arrayContaining(expectedZipFiles));
  });

  describe('apps packaged by rspack', () => {
    test('App1 Lambda Function', async () => {
      await validateLambdaFunction(
        'App1',
        cloudformation,
        testArtifactPath,
        tempAssertDir
      );
    });

    test('app3 Lambda Function', async () => {
      await validateLambdaFunction(
        'app3',
        cloudformation,
        testArtifactPath,
        tempAssertDir
      );
      const extractPath = path.join(tempAssertDir, 'app3');
      const nodeModules = fs
        .readdirSync(path.join(extractPath, 'node_modules'))
        .toString();

      expect(nodeModules).toEqual(expect.stringContaining('sharp'));
    });

    test('app4 Lambda Function', async () => {
      await validateLambdaFunction(
        'app4',
        cloudformation,
        testArtifactPath,
        tempAssertDir
      );
    });

    test('app5 Lambda Function', async () => {
      await validateLambdaFunction(
        'app5',
        cloudformation,
        testArtifactPath,
        tempAssertDir
      );
    });
  });
});
