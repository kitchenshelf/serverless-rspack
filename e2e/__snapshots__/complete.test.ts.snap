// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lambda Function Deployments Top level cloudformation 1`] = `"2010-09-09"`;

exports[`Lambda Function Deployments apps packaged by rspack App1 Lambda Function 1`] = `
{
  "DependsOn": [
    "App1LogGroup",
  ],
  "Properties": {
    "Code": {
      "S3Bucket": {
        "Ref": "ServerlessDeploymentBucket",
      },
      "S3Key": StringContaining "App1.zip",
    },
    "FunctionName": "complete-example-dev-App1",
    "Handler": "app1.handler",
    "MemorySize": 1024,
    "Role": {
      "Fn::GetAtt": [
        "IamRoleLambdaExecution",
        "Arn",
      ],
    },
    "Runtime": "provided.al2023",
    "Timeout": 6,
  },
  "Type": "AWS::Lambda::Function",
}
`;

exports[`Lambda Function Deployments apps packaged by rspack App1 Lambda Function 2`] = `
{
  "Description": "Current Lambda function version",
  "Export": {
    "Name": "sls-complete-example-dev-App1LambdaFunctionQualifiedArn",
  },
  "Value": {
    "Ref": StringContaining "App1LambdaVersion",
  },
}
`;

exports[`Lambda Function Deployments apps packaged by rspack App1 Lambda Function 3`] = `
"import{createRequire as e}from"node:module";var r={};r.n=function(e){var n=e&&e.__esModule?function(){return e.default}:function(){return e};return r.d(n,{a:n}),n},r.d=function(e,n){for(var t in n)r.o(n,t)&&!r.o(e,t)&&Object.defineProperty(e,t,{enumerable:!0,get:n[t]})},r.o=function(e,r){return Object.prototype.hasOwnProperty.call(e,r)};var n={};r.d(n,{handler:()=>o});let t=e(import.meta.url)("isin-validator");var a=r.n(t);async function o(e){return{statusCode:200,body:JSON.stringify({message:a()(e)?"ISIN is invalid!":"ISIN is fine!",input:e})}}var i=n.handler;export{i as handler};
//# sourceMappingURL=app1.mjs.map"
`;

exports[`Lambda Function Deployments apps packaged by rspack app3 Lambda Function 1`] = `
{
  "DependsOn": [
    "App3LogGroup",
  ],
  "Properties": {
    "Code": {
      "S3Bucket": {
        "Ref": "ServerlessDeploymentBucket",
      },
      "S3Key": StringContaining "app3.zip",
    },
    "FunctionName": "complete-example-dev-app3",
    "Handler": "src/App3.handler",
    "MemorySize": 1024,
    "Role": {
      "Fn::GetAtt": [
        "IamRoleLambdaExecution",
        "Arn",
      ],
    },
    "Runtime": "nodejs20.x",
    "Timeout": 6,
  },
  "Type": "AWS::Lambda::Function",
}
`;

exports[`Lambda Function Deployments apps packaged by rspack app3 Lambda Function 2`] = `
{
  "Description": "Current Lambda function version",
  "Export": {
    "Name": "sls-complete-example-dev-App3LambdaFunctionQualifiedArn",
  },
  "Value": {
    "Ref": StringContaining "App3LambdaVersion",
  },
}
`;

exports[`Lambda Function Deployments apps packaged by rspack app3 Lambda Function 3`] = `
"import{fileURLToPath as e}from"node:url";import{dirname as r}from"node:path";import{createRequire as t}from"node:module";var a={};a.n=function(e){var r=e&&e.__esModule?function(){return e.default}:function(){return e};return a.d(r,{a:r}),r},a.d=function(e,r){for(var t in r)a.o(r,t)&&!a.o(e,t)&&Object.defineProperty(e,t,{enumerable:!0,get:r[t]})},a.o=function(e,r){return Object.prototype.hasOwnProperty.call(e,r)};var n={};a.d(n,{handler:()=>f});let o=t(import.meta.url)("isin-validator");var i=a.n(o);let l=t(import.meta.url)("node:fs"),u=t(import.meta.url)("node:path");var d=a.n(u);let m=t(import.meta.url)("sharp");var p=a.n(m),s=r(e(import.meta.url));async function f(e){let r=i()(e.isin),t=d().join(s,"../my-image.jpeg"),a=(0,l.readFileSync)(t),{info:n}=await p()(a).raw().toBuffer({resolveWithObject:!0});return{statusCode:r?200:400,body:JSON.stringify({handler:"App3",message:r?"ISIN is invalid!":"ISIN is fine!",input:e,info:n})}}var c=n.handler;export{c as handler};
//# sourceMappingURL=App3.mjs.map"
`;

exports[`Lambda Function Deployments apps packaged by rspack app4 Lambda Function 1`] = `
{
  "DependsOn": [
    "App4LogGroup",
  ],
  "Properties": {
    "Code": {
      "S3Bucket": {
        "Ref": "ServerlessDeploymentBucket",
      },
      "S3Key": StringContaining "app4.zip",
    },
    "FunctionName": "complete-example-dev-app4",
    "Handler": "src/app4.handler",
    "MemorySize": 1024,
    "Role": {
      "Fn::GetAtt": [
        "IamRoleLambdaExecution",
        "Arn",
      ],
    },
    "Runtime": "nodejs20.x",
    "Timeout": 6,
  },
  "Type": "AWS::Lambda::Function",
}
`;

exports[`Lambda Function Deployments apps packaged by rspack app4 Lambda Function 2`] = `
{
  "Description": "Current Lambda function version",
  "Export": {
    "Name": "sls-complete-example-dev-App4LambdaFunctionQualifiedArn",
  },
  "Value": {
    "Ref": StringContaining "App4LambdaVersion",
  },
}
`;

exports[`Lambda Function Deployments apps packaged by rspack app4 Lambda Function 3`] = `
"import{createRequire as e}from"node:module";var r={};r.n=function(e){var n=e&&e.__esModule?function(){return e.default}:function(){return e};return r.d(n,{a:n}),n},r.d=function(e,n){for(var t in n)r.o(n,t)&&!r.o(e,t)&&Object.defineProperty(e,t,{enumerable:!0,get:n[t]})},r.o=function(e,r){return Object.prototype.hasOwnProperty.call(e,r)};var n={};r.d(n,{handler:()=>o});let t=e(import.meta.url)("isin-validator");var a=r.n(t);async function o(e){return{statusCode:200,body:JSON.stringify({message:a()(e)?"ISIN is invalid!":"ISIN is fine!",input:e})}}var i=n.handler;export{i as handler};
//# sourceMappingURL=app4.mjs.map"
`;

exports[`Lambda Function Deployments apps packaged by rspack app5 Lambda Function 1`] = `
{
  "DependsOn": [
    "App5LogGroup",
  ],
  "Properties": {
    "Code": {
      "S3Bucket": {
        "Ref": "ServerlessDeploymentBucket",
      },
      "S3Key": StringContaining "app5.zip",
    },
    "FunctionName": "complete-example-dev-app5",
    "Handler": "src/deeply/nested/somewhat/app5.handler",
    "MemorySize": 1024,
    "Role": {
      "Fn::GetAtt": [
        "IamRoleLambdaExecution",
        "Arn",
      ],
    },
    "Runtime": "nodejs20.x",
    "Timeout": 6,
  },
  "Type": "AWS::Lambda::Function",
}
`;

exports[`Lambda Function Deployments apps packaged by rspack app5 Lambda Function 2`] = `
{
  "Description": "Current Lambda function version",
  "Export": {
    "Name": "sls-complete-example-dev-App5LambdaFunctionQualifiedArn",
  },
  "Value": {
    "Ref": StringContaining "App5LambdaVersion",
  },
}
`;

exports[`Lambda Function Deployments apps packaged by rspack app5 Lambda Function 3`] = `
"import{createRequire as e}from"node:module";var r={};r.n=function(e){var n=e&&e.__esModule?function(){return e.default}:function(){return e};return r.d(n,{a:n}),n},r.d=function(e,n){for(var t in n)r.o(n,t)&&!r.o(e,t)&&Object.defineProperty(e,t,{enumerable:!0,get:n[t]})},r.o=function(e,r){return Object.prototype.hasOwnProperty.call(e,r)};var n={};r.d(n,{handler:()=>o});let t=e(import.meta.url)("isin-validator");var a=r.n(t);async function o(e){return{statusCode:200,body:JSON.stringify({message:a()(e)?"ISIN is invalid!":"ISIN is fine!",input:e})}}var i=n.handler;export{i as handler};
//# sourceMappingURL=app5.mjs.map"
`;
